docs: clean up access story (#22246)
juju/juju: docs/howto/manage-controllers.md

See more: {ref}`command-juju-bootstrap`, {ref}`cloud-specific reference docs <list-of-supported-clouds>`, {ref}`list-of-constraints`, {ref}`list-of-controller-configuration-keys`
```

`````{dropdown} Common bootstrap settings

````{dropdown} Enable external authentication

To enable external authentication via an identity provider (OAuth/OIDC), bootstrap with the `identity-url` configuration key:

```text
juju bootstrap aws my-controller --config identity-url=https://identity.example.com
```

> This changes how users authenticate:

Reviewer: It optionally changes how users authenticate.
You can still add local users. It just gives the possibility to also use externally authenticated users.

Author: Thanks for the clarification!

- Users authenticate via the external provider (e.g., Ubuntu SSO, Google OAuth) instead of controller-local passwords.
- The login process opens a browser for OAuth/OIDC authentication.
- Users can still register with tokens from `juju add-user`.

> For HTTP identity managers, you must also provide the identity manager's public key:

Reviewer: No. If the provider has the necessary endpoint to allow the public key to be retrieved, you don't need this. The candid provider `https://api.jujucharms.com/identity` for example does this.

Author: Changed to "you may need".

```text
juju bootstrap aws my-controller \
  --config identity-url=http://identity.example.com \
  --config identity-public-key=<base64-encoded-key>
```

```{ibnote}
See more: {ref}`controller-config-identity-url`, {ref}`controller-config-identity-public-key`
```

````

````{dropdown} Use JWT-based authentication

For JWT-based authentication (alternative to OAuth/OIDC or local passwords), configure the JWT token refresh URL:

```text
juju bootstrap aws my-controller --config login-token-refresh-url=https://auth.example.com/.well-known/jwks.json
```

```{ibnote}

> See more: {ref}`controller-config-login-token-refresh-url`

Reviewer: This is just used when a controller is being set up behind a JAAS front end.
I wouldn't document this as what we do is MVP and we need to refine it.

Author: Removed, thanks!

```

````

````{dropdown} Enable DNS-based registration

> To allow users to register using a DNS hostname (`juju register controller.example.com`) instead of tokens, configure the controller's DNS address:

Reviewer: Don't need the "instead of tokens"

Author: Removed, thanks!

```text
juju bootstrap aws my-controller --config public-dns-address=controller.example.com:443
```

This is typically used with external authentication:

```text
juju bootstrap aws my-controller \
  --config identity-url=https://identity.example.com \
  --config public-dns-address=controller.example.com:443
```

```{ibnote}
See more: {ref}`controller-config-public-dns-address`
```

````

````{dropdown} Enable automatic TLS certificates

To automatically obtain TLS certificates from Let's Encrypt, configure the DNS name:

```text

> juju bootstrap aws my-controller --config autocert-dns-name=controller.example.com

Reviewer: I wonder if we need to test all this. It was done so long ago, I have no confidence that it still works.

Author: Did some digging around with Claude. The autocert functionality does have unit test coverage in `internal/worker/httpserver/tls_state_test.go` (the `TLSStateAutocertSuite` tests autocert certificate retrieval, host policy, and ALPN protocol support). However, you're right that end-to-end integration testing with a real ACME server would be valuable to verify the feature works in practice. The unit tests use a mock server rather than testing against actual Let's Encrypt. Would you like me to add a note to the docs recommending users verify the feature in their environment first, or leave it as-is given the unit test coverage?

Reviewer: unit test coverage is not sufficient - I'd be reluctant to produce doc for something we're not sure works as it says on the tin.
The whole DNS based workflow in this md file (I think) should be verified by hand to validate the doc is accurate and the feature still works.

Reviewer 2: We need CI tests for that if there isn't any.

```

This is typically used with DNS addressing:

```text
juju bootstrap aws my-controller \
  --config autocert-dns-name=controller.example.com \
  --config public-dns-address=controller.example.com:443
```

```{ibnote}
See more: {ref}`controller-config-autocert-dns-name`, {ref}`controller-config-public-dns-address`
```

````

````{dropdown} Use a specific container image repository (Kubernetes)

The Juju controller needs two container images (one for the controller agent container and one for the database container). These are by default downloaded from Docker Hub, but can also be downloaded from `public.ecr.aws/juju` or `https://ghcr.io/juju`. We currently recommend `public.ecr.aws/juju`:

```text
juju bootstrap mycloud --config caas-image-repo="public.ecr.aws/juju"
```

```{note}

> While `caas-image-repo` can technically be changed after bootstrap, that is only for a very specific use case (adjusting credentials for a custom registry). For most cases it can only be set during bootstrap.

Reviewer: I did think this was read only. I am pretty sure it errors if you try and change it.

Author: You're right. Fixed it now, thanks!

```

```{ibnote}
See more: {ref}`controller-config-caas-image-repo`
```

````

````{dropdown} Set recommended controller sizing

To manage hundreds of units effectively, bootstrap with minimum 50 GB disk, 2 CPUs, and 4 GB RAM:

```text
juju bootstrap aws/us-east-1 my-controller --bootstrap-constraints "root-disk=50G cores=2 mem=4G"
```

For larger deployments, consider making the controller highly available.

```{ibnote}
See more: {ref}`manage-constraints-for-a-controller`, {ref}`make-a-controller-highly-available`
```

````

````{dropdown} Work around Kubernetes controller limitations

Kubernetes controllers currently don't support high-availability or backup/restore. For production deployments, bootstrap your controller on a machine cloud, then add your Kubernetes cloud(s) to it:

```text
juju bootstrap aws my-controller

> juju add-k8s my-k8s-cloud --controller my-controller

Reviewer: Did we want to add a note about how under the covers this will read your local kube config file (and uses the standard k8s search path, including the KUBECONFIG env var etc etc)? Or too much detail for here?

Author: Probably too much detail here, but added it anyway.

```

```{ibnote}
See more: {ref}`add-a-cloud`
```
````

````{dropdown} Customize network ports

To customize the ports used by the controller, set them during bootstrap. These cannot be changed afterward.

> **API port** (`api-port`, default: 17070):

Reviewer: Note: eventually we will move these to the controller charm, but not now

```text
juju bootstrap aws my-controller --config api-port=17080
```

**SSH server port** (`ssh-server-port`, default: 17022):

```text
juju bootstrap aws my-controller --config ssh-server-port=17023
```

**MongoDB port** (`state-port`, default: 37017):

```text
juju bootstrap aws my-controller --config state-port=37018
```

```{ibnote}
See more: {ref}`controller-config-api-port`, {ref}`controller-config-ssh-server-port`, {ref}`controller-config-state-port`
```

````

> ````{dropdown} Use a different TLS certificate provider

Reviewer: I'd love to see this all tested first before we document it

Author: See https://github.com/juju/juju/pull/22246#discussion_r3273498925

By default, when using automatic TLS certificates (`autocert-dns-name`), certificates are obtained from Let's Encrypt. To use a different ACME server (e.g., for testing), set `autocert-url` during bootstrap:

```text
juju bootstrap aws my-controller \
  --config autocert-dns-name=controller.example.com \
  --config autocert-url=https://acme-staging-v02.api.letsencrypt.org/directory
```

```{ibnote}
See more: {ref}`controller-config-autocert-url`, {ref}`controller-config-autocert-dns-name`
```

````

````{dropdown} Allow model access without controller access

> By default, users must have controller access to connect to models. To allow users to access models they're authorized for without controller access, enable `allow-model-access` during bootstrap:

Reviewer: "controller access" is a bit vague?
The lowest level is "login" access. Then "superuser".
Not sure if we can rephrase?

Author: I've investigated this a bit more (because it was a complete mystery to me). I've now added a lot more detail including an example workflow. Down the line we'll want to beef up the config reference so the story can be deduced from it -- as it is I had to ask Claude to look at the codebase.

```text
juju bootstrap aws my-controller --config allow-model-access=true
```

```{ibnote}
See more: {ref}`controller-config-allow-model-access`
```

````

````{dropdown} Use custom installation sources

**MongoDB snap channel** (for Ubuntu Focal or later, default: 4.4.30/stable):

```text

> juju bootstrap aws my-controller --config juju-db-snap-channel=5.0/stable

Reviewer: 5.0 will never exist.
Not sure if we want a better example? "4.4.31/stable"?

Author: Fixed, thanks! Also added a note for forward merge.

```

**Controller snap source** (default: legacy):

```text

> juju bootstrap aws my-controller --config jujud-controller-snap-source=snapstore

Reviewer: There is no controller snap yet. This config should not have landed I don't think.

Author: Removed.

See more: {ref}`command-juju-bootstrap`, {ref}`cloud-specific reference docs <list-of-supported-clouds>`, {ref}`list-of-constraints`, {ref}`list-of-controller-configuration-keys`
```

`````{dropdown} Common bootstrap settings

````{dropdown} Enable DNS-based registration

```{tip}

> Use DNS-based registration when: you want to simplify user onboarding (no registration tokens), you're setting up a shared controller for many users, you need automatic TLS certificates, or you want flexibility to change controller IPs without affecting users.

Reviewer 2: **suggestion:** I personaly love bullet point for this kind of stuff

```suggestion
Use DNS-based registration in any of those usecases: 
- you want to simplify user onboarding (no registration tokens), 
- you are setting up a shared controller for many users, 
- you need automatic TLS certificates, 
- you want flexibility to change controller IPs without affecting users.
```

```

````

````{dropdown} Enable automatic TLS certificates

To automatically obtain TLS certificates from Let's Encrypt, configure the DNS name:

```text
juju bootstrap aws my-controller --config autocert-dns-name=controller.example.com
```

This is typically used with DNS addressing:

```text
juju bootstrap aws my-controller \
  --config autocert-dns-name=controller.example.com \
  --config public-dns-address=controller.example.com:443
```

```{ibnote}
See more: {ref}`controller-config-autocert-dns-name`, {ref}`controller-config-public-dns-address`
```

````

````{dropdown} Use a specific container image repository (Kubernetes)

The Juju controller needs two container images (one for the controller agent container and one for the database container). These are by default downloaded from Docker Hub, but can also be downloaded from `public.ecr.aws/juju` or `https://ghcr.io/juju`. We currently recommend `public.ecr.aws/juju`:

```text
juju bootstrap mycloud --config caas-image-repo="public.ecr.aws/juju"
```

```{ibnote}
See more: {ref}`controller-config-caas-image-repo`
```

````

````{dropdown} Set recommended controller sizing

> To manage hundreds of units effectively, bootstrap with minimum 50 GB disk, 2 CPUs, and 4 GB RAM:

Reviewer: Is 4GB enough? We mandate 3.5GB just for any controller - I would have thought maybe 8GB for larger scenarios?

Author: Ian says we could maybe recommend 16-32. It depends on how many models and units you have.

```text
juju bootstrap aws/us-east-1 my-controller --bootstrap-constraints "root-disk=50G cores=2 mem=4G"
```

For larger deployments, consider making the controller highly available.

```{ibnote}
See more: {ref}`manage-constraints-for-a-controller`, {ref}`make-a-controller-highly-available`
```

````

````{dropdown} Work around Kubernetes controller limitations

> Kubernetes controllers currently don't support high-availability or backup/restore. For production deployments, bootstrap your controller on a machine cloud, then add your Kubernetes cloud(s) to it:

Reviewer 2: **question:** this target 3.6, so it is fine, but to me 4.0 does support k8s ha, isn't it ?
